tools/oeedger8r/Emitter.ml

```diff
@@ -154,42 +154,41 @@ let attr_value_to_string argstruct = function
 (** For a parameter, get its size expression. *)
 let oe_get_param_size (ptype, decl, argstruct) =
   let type_expr = get_type_expr ptype in
-  (* TODO: Handle edge case of [size * count]. *)
-  let get_ptr_or_decl_size (p : ptr_size) (d : declarator) =
-    match attr_value_to_string argstruct p.ps_count with
-    | Some s -> sprintf "(%s * sizeof(%s))" s type_expr
-    | None -> (
-      match attr_value_to_string argstruct p.ps_size with
-      | Some s -> s
-      | None ->
-          let dims =
-            List.map (fun i -> "[" ^ string_of_int i ^ "]") d.array_dims
-          in
-          sprintf "sizeof(%s%s)" type_expr (String.concat "" dims) )
+  let get_ptr_or_decl_size (p : ptr_size) =
+    let size = attr_value_to_string argstruct p.ps_size
+    and count = attr_value_to_string argstruct p.ps_count in
+    match (size, count) with
+    | Some s, None -> s
+    | None, Some c -> sprintf "(%s * sizeof(%s))" c type_expr
+    (* TODO: Check that this is an even multiple of the size of type. *)
+    | Some s, Some c -> sprintf "(%s * %s)" s c
+    | None, None ->
+        sprintf "sizeof(%s%s)" type_expr (get_array_dims decl.array_dims)
   in
   match ptype with
   | PTPtr (_, ptr_attr) ->
       if ptr_attr.pa_isstr then
         argstruct ^ decl.identifier ^ "_len * sizeof(char)"
       else if ptr_attr.pa_iswstr then
         argstruct ^ decl.identifier ^ "_len * sizeof(wchar_t)"
-      else get_ptr_or_decl_size ptr_attr.pa_size decl
+      else get_ptr_or_decl_size ptr_attr.pa_size
   (* Values have no marshalling size. *)
   | _ -> ""
 
 (** For a parameter, get its count expression. *)
 let oe_get_param_count (ptype, decl, argstruct) =
   let type_expr = get_type_expr ptype in
-  (* TODO: Handle edge case of [size * count]. *)
-  let get_ptr_or_decl_count (p : ptr_size) (d : declarator) =
-    match attr_value_to_string argstruct p.ps_count with
-    | Some s -> s
-    | None -> (
-      match attr_value_to_string argstruct p.ps_size with
-      | Some s -> sprintf "(%s / sizeof(%s))" s type_expr
-      | None ->
-          let dims = List.map string_of_int d.array_dims in
-          String.concat " * " dims )
+  let get_ptr_or_decl_count (p : ptr_size) =
+    let size = attr_value_to_string argstruct p.ps_size
+    and count = attr_value_to_string argstruct p.ps_count in
+    match (size, count) with
+    | Some s, None -> sprintf "(%s / sizeof(%s))" s type_expr
+    | None, Some c -> c
+    (* TODO: Check that this is an even multiple of the size of type. *)
+    | Some s, Some c -> sprintf "((%s * %s) / sizeof(%s))" s c type_expr
+    | None, None ->
+        let dims = List.map string_of_int decl.array_dims in
```

Comment:
Isn't this just get_array_dims like you changed above?

Reply:
Whoops, that's what happens when I have multiple PRs in progress and have to reorder them 😅

Reply:
Wait nope, this is doing something entirely different than get_array_dims. This is giving us the expression `2 * 4 * 3 * 5` etc., instead of `[2][4][3][5]`.